README.md
# Rampant Maintenance

Machines will periodically breakdown (become inactive for cooldown and take some damage). When damaged machines breakdown more damage is taken and more time is required to come back up. Tech to reduce breakdowns.  

# Forum Post


# Notes

# Features

- Machines will periodically breakdown (become inactive for cooldown and take some damage)  
  - Inserters
  - Labs
  - Crafting Machines
  - Mining Drill
  - Rail Signals*
  - Chain Signals*
  - Generators
  - Boilers
  - Reactors
  - Centrifuge
  - Turrets
  - Lamps
  - Radar
  - Offshore Pump
  - Burner Generator
  - Pumps
  - Oil Refinery
  - Chemical Lab
  - Artillery Turret
  - Pumpjack
  - Solar Panel
  - Accumulator
  - Turbine
  - Heat Exchanger
  - Beacon
  - Roboport
- When damaged machines breakdown more damage is taken and more time is required to come back up  
- Tech to reduce breakdowns  
- Idle Machines have a greatly reduced chance of breakdown  

# Planned Features


changelog.txt
---------------------------------------------------------------------------------------------------
Version: 1.4.2
  Bugfixes:
    - Fixed player repaired structures not correctly setting broken record flag
    - Fixed reactor breaking down when not working

---------------------------------------------------------------------------------------------------
Version: 1.4.1
  Tweaks:
    - Changed mod setting that robot repairs reactive entities to be off by default

---------------------------------------------------------------------------------------------------
Version: 1.4.0
  Compatibility:
    - Added support for krastorio reinforced plates as a tile with 35% bonus
    - Added remote interface to add and remove entities from breakdowns
  Improvements:
    - Added active flag to separate when this mod deactivates or activates an entity vs another mod
  Tweaks:
    - Changed unknown tile type to have 0% modifier
    - onModSettingsChange all internal entity state is synced

---------------------------------------------------------------------------------------------------
Version: 1.3.0
  Improvements:
    - Added mod setting to enable robot repairs re-activating machines
  Bugfixes:
    - Fixed on upgrade breaking gui toggle for existing guis
    - Fixed player leaving causing crash with guis

---------------------------------------------------------------------------------------------------
Version: 1.2.2
  Bugfixes:
    - Fixed nil value tilePrototype due to script raised_set_tiles
    - Fixed during multiplayer per player gui should correctly fix up iterator when player leaves

---------------------------------------------------------------------------------------------------
Version: 1.2.1
Date: 20. 03. 2022
  Tweaks:
    - Changed the license to GPLv3
  Bugfixes:
    - Corrected tooltip on shortcut bar

---------------------------------------------------------------------------------------------------
Version: 1.2.0
Date: 01. 03. 2022
  Improvements:
    - Modified RND to be rolled per entity checked instead of for a set of entities
    - Added MTBF, MTTR, Active, Inactive, Uptime, Failure Count
    - Added popup metrics when hovering entities that can breakdown
    - Added tile modifier and pollution modifier to modify breakdowns, damage, and downtime
    - Added tile and pollution modifier research
    - Added per entity type mod setting tile base modifier
    - Added per entity type mod setting pollution max
  Tweaks:
    - Reduced damage research reduction by 50% per research level
    - Increased default accumulator failure-rate by 0.1 and increased damage by 2x
    - Increased default ammo-turret failure-rate by 0.2 and increased downtime by 2 minutes and increased damage by ~1.5x
    - Increased default artillery-turret failure-rate by 0.2 and increased downtime by 2 minutes and increased damage by ~2x
    - Increased default assembling-machine failure-rate by 0.1 and increased downtime by ~2 minutes and increased damage by ~2x
    - Increased default beacon damage by ~0.5x
    - Increased default boiler failure-rate by 0.1 and increased damage by ~2x
    - Increased default electric-pole damage by ~0.5x
    - Increased default electric-turret failure-rate by 0.1 and increased downtime by ~3 minutes and increased damage by ~0.3x
    - Increased default fluid-turret failure-rate by 0.1 and increased downtime by ~3 minutes and increased damage by ~2x
    - Increased default furnace failure-rate by 0.1 and increased downtime by ~1 minute
    - Increased default generator failure-rate by 0.1 and increased damage by ~2x
    - Increased default inserter failure-rate by 0.1 and increased damage by ~0.5x
    - Increased default lab damage by ~0.5x
    - Increased default lamp damage by ~0.5x
    - Increased default mining drill failure-rate by 0.1 and increased downtime by ~2 and increased damage by ~2x
    - Increased default offshore-pump failure-rate by 0.1 and increased damage by ~2x
    - Increased default pump failure-rate by 0.1 and increased damage by ~2x
    - Increased default radar failure-rate by 0.1 and increased damage by ~0.5x
    - Increased default reactor failure-rate by 0.2 and increased downtime by 2 minutes and increased damage by ~2x
    - Increased default roboport failure-rate by 0.1 and increased damage by ~2x
    - Increased default rocket-port failure-rate by 0.1 and increased damage by ~2.5x
    - Increased default solar panel failure-rate by 0.1 and increased damage by ~0.5x
  Bugfixes:
    - Entities that didn't have a predicate check would not be cleared

---------------------------------------------------------------------------------------------------
Version: 1.1.0
Date: 27. 8. 2021
  Improvements:
    - When the player fully repairs a broke down machine, it will be reactived. Currently the brokedown icon will remain until the machine would have restarted naturally. Robots repairing a deactivated building do not trigger this effect.

---------------------------------------------------------------------------------------------------
Version: 1.0.4
Date: 16. 4. 2021
  Bugfixes:
    - Fixed mod settings being refreshed on all mod changes

---------------------------------------------------------------------------------------------------
Version: 1.0.3
Date: 06. 2. 2021
  Tweaks:
    - Decreased default damage taken on breakdown for non-turrets to 15%
    - Reduced breakdown icon size and moved to corner of entity
    - Increased default min downtime to 500 seconds and max downtime to 800 seconds
    - Decreased all research efficiency gains to 8%
    - Increased cooldowns for roughly all entities by 2 minutes
    - Reorganized map settings
    - Increased failure rate for everything except electric poles to 10% min and 20% max
  Bugfixes:
    - Fixed breakdown logic calculating incorrect future tick events
    - Fixed percentage in technology descriptions
    - Fixed artillery turret predicate for breakdown


libs/Constants.lua
-- Copyright (C) 2022  veden

-- This program is free software: you can redistribute it and/or modify
-- it under the terms of the GNU General Public License as published by
-- the Free Software Foundation, either version 3 of the License, or
-- (at your option) any later version.

-- This program is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU General Public License for more details.

-- You should have received a copy of the GNU General Public License
-- along with this program.  If not, see <https://www.gnu.org/licenses/>.


if constantsG then
    return constantsG
end

local mMin = math.min
local mMax = math.max

local constants = {}

constants.ALL_ENTITY_TYPES = {
    "accumulator", "ammo-turret", "artillery-turret", "assembling-machine", "beacon",
    "boiler", "electric-pole", "electric-turret", "fluid-turret", "furnace", "generator",
    "inserter", "lab", "lamp", "mining-drill", "offshore-pump", "pump", "radar", "reactor",
    "roboport", "rocket-silo", "solar-panel"
}
constants.ENTITES_WITHOUT_DOWNTIME = {
    ["solar-panel"] = true,
    ["lamp"] = true,
    ["electric-pole"] = true,
    ["accumulator"] = true
}

constants.RESEARCH_LOOKUP = {}
for i=1,9 do
    constants.RESEARCH_LOOKUP["rampant-maintenance-reduce-failure-"..i] = "failure"
    constants.RESEARCH_LOOKUP["rampant-maintenance-reduce-damage-"..i] = "damage"
    constants.RESEARCH_LOOKUP["rampant-maintenance-reduce-damage-failure-"..i] = "damage-failure"
    constants.RESEARCH_LOOKUP["rampant-maintenance-reduce-downtime-"..i] = "downtime"
    constants.RESEARCH_LOOKUP["rampant-maintenance-reduce-checks-"..i] = "cooldown"
    constants.RESEARCH_LOOKUP["rampant-maintenance-reduce-tile-"..i] = "tile"
    constants.RESEARCH_LOOKUP["rampant-maintenance-reduce-pollution-"..i] = "pollution"
    constants.RESEARCH_LOOKUP["rampant-maintenance-reduce-energy-"..i] = "energy"
end

constants.POPUP_TTL = 60 * 5
constants.TICKS_PER_MINUTE = 60 * 60
constants.TICKS_PER_HOUR = constants.TICKS_PER_MINUTE * 60
constants.TICKS_PER_FIVE_HOURS = constants.TICKS_PER_MINUTE * 60 * 5

constants.TERRAIN_MODIFIERS = {
    {"reinforced", -0.35},
    {"refined", -0.3},
    {"marking", -0.3},
    {"asphalt", -0.3},
    {"concrete", -0.2},
    {"stone", -0.1},
    {"grass", 0},
    {"landfill", 0.1},
    {"dirt", 0.1},
    {"snow", 0.2},
    {"volcanic", 0.2},
    {"desert", 0.3},
    {"sand", 0.3},
    {"nuclear", 0.3}
}

constants.POLLUTION_TO_PERCENTAGE = 1 / 2000

function constants.roundToNearest(number, multiple)
    local num = number + (multiple * 0.5)
    return num - (num % multiple)
end

function constants.getResearch(forceName, world, typeName)
    local researches = world.forceResearched[forceName]
    if typeName == "tile" or typeName == "energy" then
        return (researches and researches[typeName]) or 0
    else
        return (researches and researches[typeName]) or 1
    end
end

function constants.calculateLowFailure(world, entityRecord, invertedHealthPercent)
    local entity = entityRecord.e
    local entityType = entity.type
    local entityForceName = entity.force.name
    local buildFailure = world.buildFailure[entityType]
    local buildDamageFailure = world.buildDamageFailure[entityType]
    local failureChance = (buildFailure.low * constants.getResearch(entityForceName, world, "failure"))
    local damageFailureChance = (buildDamageFailure.low * invertedHealthPercent) * constants.getResearch(entityForceName, world, "damage-failure")
    local pollutionModifier = (1 + (mMin(entity.surface.get_pollution(entity.position) * constants.POLLUTION_TO_PERCENTAGE,
                                         entityRecord.p * constants.getResearch(entityForceName, world, "pollution"))))
    local tileModifier = (1 + (entityRecord.t + entityRecord.tM + constants.getResearch(entityForceName, world, "tile")))
    return (failureChance + damageFailureChance) * pollutionModifier * tileModifier
end

function constants.calculateHighFailure(world, entityRecord, invertedHealthPercent)
    local entity = entityRecord.e
    local entityType = entity.type
    local entityForceName = entity.force.name
    local buildFailure = world.buildFailure[entityType]
    local buildDamageFailure = world.buildDamageFailure[entityType]
    local failureChance = ((buildFailure.low + buildFailure.range) * constants.getResearch(entityForceName, world, "failure"))
    local damageFailureChance = ((buildDamageFailure.low + buildDamageFailure.range) * invertedHealthPercent) * constants.getResearch(entityForceName, world, "damage-failure")
    local pollutionModifier = (1 + (mMin(entity.surface.get_pollution(entity.position) * constants.POLLUTION_TO_PERCENTAGE,
                                         entityRecord.p * constants.getResearch(entityForceName, world, "pollution"))))
    local tileModifier = (1 + (entityRecord.t + entityRecord.tM + constants.getResearch(entityForceName, world, "tile")))
    return (failureChance + damageFailureChance) * pollutionModifier * tileModifier
end

function constants.calculateLowCooldown(world, entityRecord, healthPercent)
    local entity = entityRecord.e
    local cooldowns = world.buildCooldown[entity.type]
    local cooldown = mMax(0.2, healthPercent) * cooldowns.low
    return cooldown * (1 + constants.getResearch(entity.force.name, world, "cooldown"))
end

function constants.calculateHighCooldown(world, entityRecord, healthPercent)
    local entity = entityRecord.e
    local cooldowns = world.buildCooldown[entity.type]
    local cooldown = mMax(0.2, healthPercent) * (cooldowns.low + cooldowns.range)
    return cooldown * (1 + constants.getResearch(entity.force.name, world, "cooldown"))
end

function constants.calculateLowDowntime(world, entityRecord, invertedHealthPercent)
    local entity = entityRecord.e
    local entityForceName = entity.force.name
    local downtimes = world.buildDowntime[entity.type]
    if not downtimes then
        return 0
    end
    local downtime = (downtimes.low * mMax(1, 1+invertedHealthPercent)) * constants.getResearch(entityForceName, world, "downtime")
    local tileModifier = (1 + (entityRecord.t + entityRecord.tM + constants.getResearch(entityForceName, world, "tile")))
    local pollutionModifier = (1 + (mMin(entity.surface.get_pollution(entity.position) * constants.POLLUTION_TO_PERCENTAGE,
                                         entityRecord.p * constants.getResearch(entityForceName, world, "pollution"))))
    return downtime * tileModifier * pollutionModifier
end

function constants.calculateHighDowntime(world, entityRecord, invertedHealthPercent)
    local entity = entityRecord.e
    local entityForceName = entity.force.name
    local downtimes = world.buildDowntime[entity.type]
    if not downtimes then
        return 0
    end
    local downtime = ((downtimes.low + downtimes.range) * mMax(1, 1+invertedHealthPercent)) *
        constants.getResearch(entityForceName, world, "downtime")
    local tileModifier = (1 + (entityRecord.t + entityRecord.tM + constants.getResearch(entityForceName, world, "tile")))
    local pollutionModifier = (1 + (mMin(entity.surface.get_pollution(entity.position) * constants.POLLUTION_TO_PERCENTAGE,
                                         entityRecord.p * constants.getResearch(entityForceName, world, "pollution"))))
    return downtime * tileModifier * pollutionModifier
end

function constants.calculateLowDamage(world, entityRecord, invertedHealthPercent)
    local entity = entityRecord.e
    local entityForceName = entity.force.name
    local damages = world.buildDamage[entity.type]
    local damage = damages.low * constants.getResearch(entityForceName, world, "damage")
    local tileModifier = (1 + (entityRecord.t + entityRecord.tM + constants.getResearch(entityForceName, world, "tile")))
    local pollutionModifier = (1 + (mMin(entity.surface.get_pollution(entity.position) * constants.POLLUTION_TO_PERCENTAGE,
                                         entityRecord.p * constants.getResearch(entityForceName, world, "pollution"))))
    return (damage + (damage * invertedHealthPercent)) * tileModifier * pollutionModifier
end

function constants.calculateHighDamage(world, entityRecord, invertedHealthPercent)
    local entity = entityRecord.e
    local entityForceName = entity.force.name
    local damages = world.buildDamage[entity.type]
    local damage = (damages.low + damages.range) * constants.getResearch(entityForceName, world, "damage")
    local tileModifier = (1 + (entityRecord.t + entityRecord.tM + constants.getResearch(entityForceName, world, "tile")))
    local pollutionModifier = (1 + (mMin(entity.surface.get_pollution(entity.position) * constants.POLLUTION_TO_PERCENTAGE,
                                         entityRecord.p * constants.getResearch(entityForceName, world, "pollution"))))
    return (damage + (damage * invertedHealthPercent)) * tileModifier * pollutionModifier
end

constantsG = constants
return constantsG
